fsspec/implementations/cache_metadata.py

```diff
@@ -0,0 +1,210 @@
+
+    from typing_extensions import TypeAlias
+
+    Detail: TypeAlias = Dict[str, Any]
+
+
+class CacheMetadata:
+    """Cache metadata.
+
+    All reading and writing of cache metadata is performed by this class,
+    accessing the cached files and blocks is not.
+
+    Metadata is stored in a single file per storage directory, pickled.
+    """
+
+    def __init__(self, storage: list[str]):
+        """
+
+        Parameters
+        ----------
+        storage: list[str]
+            Directories containing cached files, must be at least one. Metadata
+            is stored in the last of these directories by convention.
+        """
+        if not storage:
+            raise ValueError("CacheMetadata expects at least one storage location")
+
+        self._storage = storage
+        self.cached_files: list[Detail] = [{}]
+
+    def _scan_locations(
+        self, writable_only: bool = False
+    ) -> Iterator[tuple[str, str, bool]]:
+        """Yield locations (filenames) where metadata is stored, and whether
+        writable or not.
+
+        Parameters
+        ----------
+        writable: bool
+            Set to True to only yield writable locations.
+
+        Returns
+        -------
+        Yields (str, str, bool)
+        """
+        n = len(self._storage)
+        for i, storage in enumerate(self._storage):
+            writable = i == n - 1
+            if writable_only and not writable:
+                continue
+            yield os.path.join(storage, "cache"), storage, writable
+
+    def check_file(
+        self, path: str, ignore: Callable[[Detail], bool] | None
+    ) -> Literal[False] | tuple[Detail, str]:
+        """If path is in cache return its details, otherwise return ``False``.
+
+        Takes an optional callback so that the caller can choose to ignore
+        possible matches if, for example, they are too old.
+        """
+        for (fn, base, _), cache in zip(self._scan_locations(), self.cached_files):
+            if path not in cache:
+                continue
+            detail = cache[path].copy()
+            if ignore is not None and ignore(detail):
+                continue
+            fn = os.path.join(base, detail["fn"])
+            if os.path.exists(fn):
+                return detail, fn
+        return False
+
+    def clear_expired(self, expiry_time: int) -> Iterator[str]:
+        """Remove expired metadata from the cache.
+
+        Yields names of files corresponding to expired metadata. Caller is
+        responsible for deleting these files.
+        """
+        for path, detail in self.cached_files[-1].copy().items():
+            if time.time() - detail["time"] > expiry_time:
+                fn = detail.get("fn", "")
+                if not fn:
+                    raise RuntimeError(
+                        f"Cache metadata does not contain 'fn' for {path}"
+                    )
+                fn = os.path.join(self._storage[-1], fn)
```

Comment:
Should be "/", i.e., are we assuming the storage locations are fsspec-style URLs ?

Reply:
(also occurs elsewhere)

Reply:
It looks like historically we have always used `os.path.join` for such manipulations. Our docs and tests seem to have a mixture of fsspec-style URLs and OS-specific paths.

I'll try standardising on forward slashes and see what breaks on Windows.

Reply:
Just noticed that you thought we could defer this change, so I'll leave this until I've fixed everything else.